### PR TITLE
Upgrade preparation to verify sysctl values containing forward slashes by regex

### LINKED ISSUE #102373: Regex for validating Sysctl values in deployments might not be correct.
#### What happened:
The Regex that validates sysctl values in objects doesn't allow names that contain forward slashes. Sysctl values on network subinterfaces contain these forward slashes.

#### What you expected to happen:
Running:
kubectl create -f ./deployment.yaml

Creates the deployment.

#### How to reproduce it (as minimally and precisely as possible):

Create a deployment that brings an additional network interface with subinterface into a pod (using something like Multus). 

Interface will appear as eth2.100

sysctl values for this interface will follow the convention "net.ipv4.conf.eno2/100..."

Attempt to set one of these values in the deployment such as:

spec:
  securityContext:
    sysctls:
    - name: net.ipv4.conf.eno2/100.rp_filter
      value: "0"

It will fail validation, even though it seems like that's the valid name. I wasn't able to escape it.

#### Environment:
- Kubernetes version (use `kubectl version`): 1.20.5
- Cloud provider or hardware configuration: On-premise
- OS (e.g: `cat /etc/os-release`): Centos 7
- Kernel (e.g. `uname -a`): 5.5.2-1
- Install tools: kubeadm init
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by mengjiao-liu**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
The Regex that validates sysctl values in objects doesn't allow names that contain forward slashes. Sysctl values on network subinterfaces contain these forward slashes.

I looked up Linux's description of the sysctl variable:
- https://man7.org/linux/man-pages/man8/sysctl.8.html
- https://man7.org/linux/man-pages/man5/sysctl.d.5.html
```
       variable
              The name of a key to read from.  An example is
              kernel.ostype.  The '/' separator is also accepted in
              place of a '.'.
```
Therefore, the dots and slashes serve the same purpose, and the sysctl variable should also support slashes as separator in Kubernetes.

As https://github.com/kubernetes/kubernetes/pull/102393#discussion_r645018213, the PR does the following:
- use strict validation for Create requests, and for Update requests where the existing object's sysctl does not contain a slash
- use relaxed validation for Update requests where the existing object's sysctl contains a slash

#### Which issue(s) this PR fixes:

Ref #102373

#### Does this PR introduce a user-facing change?

```release-note
Pod SecurityContext sysctls name parameter for update requests where the existing object's sysctl contains slashes and kubelet sysctl whitelist support contains slashes.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/apis/core/validation/validation.go
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:3777]: Relaxing validation of API fields has to be done over two releases so that we don't allow data to be persisted that the prior version of the API server considered invalid. That breaks multi-server clusters that are in the middle of an upgrade and have an API server at version N that allows data to be persisted with the relaxed validation, but still has an API server at version N-1 that rejects an update to that object because it is still enforcing the stricter validation.

Comment (liggitt) [reply pkg/apis/core/validation/validation.go:3777]: In release 1:
* use strict validation for Create requests, and for Update requests where the existing object's sysctl does not contain a slash
* use relaxed validation for Update requests where the existing object's sysctl contains a slash

In release 2:
* use relaxed validation everywhere

This implies all the places we do validation know whether to apply strict or relaxed validation, and that we determine that based on existing data in Update requests. See construction of PodValidationOptions in pkg/api/pod/util.go, and propagation to all pod/podspec validation methods for a pattern for handling this

Comment (mengjiao-liu) [reply pkg/apis/core/validation/validation.go:3777]: oh I see now and it's better to deal with it this way. I'll update that, thanks!

**Comment by mengjiao-liu** [comment]
ping @liggitt

File: pkg/api/pod/util.go
Comment (liggitt) [inline pkg/api/pod/util.go:337]: rename to `hasSysctlsWithSlashNames`

File: pkg/api/pod/util.go
Comment (liggitt) [inline pkg/api/pod/util.go:341]: return early
```go
if podSpec.SecurityContext == nil {
  return false
}
```

File: pkg/api/pod/util.go
Comment (liggitt) [inline pkg/api/pod/util.go:351]: no need to check length of Sysctls or Name, simplify to
```go
for _, s := range securityContext.Sysctls {
  if strings.Contains(s.Name, "/") {
    return true
  }
}
return false
```

File: pkg/api/pod/util.go
Comment (liggitt) [inline pkg/api/pod/util.go:346]: this is incorrect, it means an empty name would return false even if later sysctls had a slash

File: pkg/api/pod/util.go
Comment (liggitt) [inline pkg/api/pod/util.go:460]: add a comment that in 1.23 we intend to relax this to be true by default

File: pkg/apis/core/validation/validation.go
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:3786]: keep the MustCompile as a package var, this is expensive to do on every validation check

File: pkg/apis/core/validation/validation.go
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:3799]: don't construct these dynamically, have two package var regexes and select between them

File: pkg/api/pod/util.go
Comment (mengjiao-liu) [reply pkg/api/pod/util.go]: OK, it's been modified. Thanks!

File: pkg/api/pod/util.go
Comment (mengjiao-liu) [reply pkg/api/pod/util.go:341]: Updated. Thanks.

File: pkg/api/pod/util.go
Comment (mengjiao-liu) [reply pkg/api/pod/util.go]: Updated.

File: pkg/api/pod/util.go
Comment (mengjiao-liu) [reply pkg/api/pod/util.go]: You're right. Updated and Thanks.

File: pkg/api/pod/util.go
Comment (mengjiao-liu) [reply pkg/api/pod/util.go]: Has been added. Thanks.

File: pkg/apis/core/validation/validation.go
Comment (mengjiao-liu) [reply pkg/apis/core/validation/validation.go:3786]: I haven't considered this before, thank you for your reminder.

File: pkg/apis/core/validation/validation.go
Comment (mengjiao-liu) [reply pkg/apis/core/validation/validation.go:3810]: Updated.

File: pkg/api/pod/util.go
Comment (liggitt) [inline pkg/api/pod/util.go:342]: can drop the length check to simplify; the range will be a no-op on a zero length Sysctls list

File: pkg/apis/core/validation/validation.go
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:3782]: this allows slashes in every segment... is that correct?

File: pkg/apis/core/validation/validation.go
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:3788]: this allows multiple slashes in a row... is that correct?

File: pkg/apis/core/validation/validation.go
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:3800]: keep the parameter and the package variable names consistent (use "relaxed" for both, or use "canContainSlash" for both)

File: pkg/apis/core/validation/validation.go
Comment (mengjiao-liu) [reply pkg/apis/core/validation/validation.go:3782]: I read the Linux [sysctl documentation](https://man7.org/linux/man-pages/man5/sysctl.d.5.html) and found that slashes  be used as separators within sysctl variable names.
```
       Note that either "/" or "."  may be used as separators within
       sysctl variable names. If the first separator is a slash,
       remaining slashes and dots are left intact. If the first
       separator is a dot, dots and slashes are interchanged.
       "kernel.domainname=foo" and "kernel/domainname=foo" are
       equivalent and will cause "foo" to be written to
       /proc/sys/kernel/domainname. Either
       "net.ipv4.conf.enp3s0/200.forwarding" or
       "net/ipv4/conf/enp3s0.200/forwarding" may be used to refer to
       /proc/sys/net/ipv4/conf/enp3s0.200/forwarding.
```
Therefore, the relaxed regex needs to be modified to:
`([a-z0-9]([-_/a-z0-9]*[a-z0-9])?[\./])*[a-z0-9]([-_a-z0-9]*[a-z0-9])?`

File: pkg/apis/core/validation/validation.go
Comment (mengjiao-liu) [reply pkg/apis/core/validation/validation.go:3788]: Explain the same as above, need to modify the regex.

File: pkg/apis/core/validation/validation.go
Comment (mengjiao-liu) [reply pkg/apis/core/validation/validation.go]: Updated.

**Comment by mengjiao-liu** [comment]
ping @liggitt

**Comment by mengjiao-liu** [comment]
Any chance you'll find time to review it? @liggitt

File: pkg/apis/core/validation/validation.go
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:3796]:
```suggestion
// i.e. matches SysctlFmt (or SysctlContainSlashFmt if canContainSlash is true).
```

**Comment by liggitt** [comment]
The mechanics for relaxing this look correct to me

However, the kubelet will still reject sysctls in this format. The kubelet only permits a subset of sysctls by default, and validates any additional permitted sysctls here:

https://github.com/kubernetes/kubernetes/blob/e974bcd471fa562a000ca9780a479c01b0dfe84f/pkg/kubelet/sysctl/whitelist.go#L51-L57

by calling this method (which still has the stricter validation):

https://github.com/kubernetes/kubernetes/blob/e974bcd471fa562a000ca9780a479c01b0dfe84f/pkg/apis/policy/validation/validation.go#L375-L381

The kubelet should switch to calling `k8s.io/kubernetes/pkg/apis/core/validation#IsValidSysctlName(..., true /* allowSlashes */)`

**Review by liggitt — COMMENTED** [review]
would like a node approver as well to ack that allowing slashes in sysctls is correct/safe (xref https://github.com/kubernetes/kubernetes/pull/102393#discussion_r654125108 for links to official docs)

**Comment by liggitt** [comment]
also, these sentences from https://github.com/kubernetes/kubernetes/pull/102393#discussion_r654125108 made my head hurt :)

> Note that either "/" or "."  may be used as separators within sysctl variable names.
> If the first separator is a slash, remaining slashes and dots are left intact.
> If the first separator is a dot, dots and slashes are interchanged.

**Comment by mengjiao-liu** [comment]
> also, these sentences from [#102393 (comment)](https://github.com/kubernetes/kubernetes/pull/102393#discussion_r654125108) made my head hurt :)
> 
> > Note that either "/" or "."  may be used as separators within sysctl variable names.
> > If the first separator is a slash, remaining slashes and dots are left intact.
> > If the first separator is a dot, dots and slashes are interchanged.

I tested it, and both of these writing methods are ok when setting or reading.
It's just that `sysctl` command will automatically convert dots and slashes when displaying, so in regular expressions, dots and slashes have the same effect. 

Example：
```
Original value
# sysctl vm.max_map_count
vm.max_map_count = 262144
# sysctl vm/max_map_count
vm.max_map_count = 262144

# sysctl vm.max_map_count=262143
vm.max_map_count = 262143

# sysctl vm/max_map_count=262142
vm.max_map_count = 262142

# sysctl vm.max_map_count
vm.max_map_count = 262142
# sysctl vm/max_map_count
vm.max_map_count = 262142
```

Both display methods point to the same file.
```
$ cat /proc/sys/vm/max_map_count
262142
```
Therefore, the regular verification here should not consider the conversion of dots and slashes, but only allow slashes to exist, just my shallow understanding.

File: pkg/apis/core/validation/validation.go
Comment (mengjiao-liu) [reply pkg/apis/core/validation/validation.go:3796]: Thanks for catch. Will be updated.

**Comment by mengjiao-liu** [comment]
from https://github.com/kubernetes/kubernetes/pull/102393#issuecomment-869082781

> The kubelet should switch to calling k8s.io/kubernetes/pkg/apis/core/validation#IsValidSysctlName(..., true /* allowSlashes */)

I don't think so. the `pkg/apis/policy/validation/validation.go  IsValidSysctlPattern` function should be modified to support relaxed  validation, because the regular expressions used by the `pkg/apis/core/validation/validation.go  IsValidSysctlName` function and the IsValidSysctlPattern function are different.

IsValidSysctlPattern use SysctlPatternFmt regex support `*` char and that regex pkg/apis/core/validation/validation.go sysctlRegexp/SysctlContainSlashFmt does not.
This regular expression supports the * wildcard and that regular expression does not

sysctl Whitelist contains SafeSysctl Whitelist and allowedUnsafeSysctls
https://github.com/kubernetes/kubernetes/blob/6010cbe593f88fbddae78e43e40842e95ba3f2a1/pkg/kubelet/kubelet.go#L792-L795
allowedUnsafeSysctls parameter supports `*`matching, see https://kubernetes.io/docs/tasks/administer-cluster/sysctl-cluster/#enabling-unsafe-sysctls
```
kubelet --allowed-unsafe-sysctls \
  'kernel.msg*,net.core.somaxconn' ...
```
so, this pkg/apis/policy/validation/validation.go SysctlPatternFmt regex (need to contain relaxed regex) should be used for verification. Right?

**Comment by liggitt** [comment]
> I am a bit confused about this sentence. the `pkg/apis/policy/validation/validation.go IsValidSysctlPattern` function should be modified to support relaxed validation, because the regular expressions used by the `pkg/apis/core/validation/validation.go IsValidSysctlName` function and the IsValidSysctlPattern function are different.

You're right... though IsValidSysctlPattern is also used in PodSecurityPolicy API validation, and we don't want to relax that validation. So I think that needs the same treatment where it accepts a boolean indicating whether a slash is allowed, and the kubelet should pass `true` and PSP API validation should pass `false`

**Comment by mengjiao-liu** [comment]
https://github.com/kubernetes/kubernetes/pull/102393#pullrequestreview-693368935
PTAL @mrunalp

File: pkg/kubelet/sysctl/whitelist.go
Comment (liggitt) [inline pkg/kubelet/sysctl/whitelist.go:54]: must match `policyvalidation.SysctlContainSlashPatternFmt`

File: pkg/kubelet/sysctl/whitelist.go
Comment (liggitt) [inline pkg/kubelet/sysctl/whitelist.go:48]: add unit tests to pkg/kubelet/sysctl/whitelist_test.go to make sure we can allow a sysctl in the format you want, and that the exact value and prefix matching work properly

**Comment by liggitt** [comment]
API bits lgtm, error message and unit tests need update, and would like lgtm from node reviewer

File: pkg/kubelet/sysctl/whitelist.go
Comment (mengjiao-liu) [reply pkg/kubelet/sysctl/whitelist.go:48]: Ok, will update as soon as possible tomorrow.

Comment (mengjiao-liu) [reply pkg/kubelet/sysctl/whitelist.go:48]: Updated.

File: pkg/kubelet/sysctl/namespace.go
Comment (liggitt) [inline pkg/kubelet/sysctl/namespace.go:53]: at first glance, this doesn't seem to follow the logic described where the first separator determines how the remaining separators are handled

File: pkg/kubelet/sysctl/whitelist.go
Comment (liggitt) [reply pkg/kubelet/sysctl/whitelist.go:54]: I mean the error message this prints needs to change to reference SysctlContainSlashPatternFmt

File: pkg/kubelet/sysctl/whitelist.go
Comment (liggitt) [inline pkg/kubelet/sysctl/whitelist.go:86]: this also doesn't seem to follow the logic described where the first separator determines how the remaining separators are handled

File: pkg/kubelet/sysctl/whitelist.go
Comment (mengjiao-liu) [reply pkg/kubelet/sysctl/whitelist.go:54]: ok, got it.

File: pkg/kubelet/sysctl/whitelist.go
Comment (mengjiao-liu) [reply pkg/kubelet/sysctl/whitelist.go:86]: Same as above.

File: pkg/kubelet/sysctl/namespace.go
Comment (mengjiao-liu) [reply pkg/kubelet/sysctl/namespace.go:53]: https://github.com/kubernetes/kubernetes/blob/4748bb04b61f8717641913a0ff3a183ad28e64c9/pkg/kubelet/sysctl/namespace.go#L37-L47

https://github.com/kubernetes/kubernetes/blob/4748bb04b61f8717641913a0ff3a183ad28e64c9/pkg/security/podsecuritypolicy/sysctl/mustmatchpatterns.go#L32-L40

The namespace listed contains only dots. If the enumeration contains sysctl for slashes and dots, that would be too many items. Any better suggestion？

**Comment by SergeyKanzhelev** [comment]
I'd suggest to add a release note to this PR

**Comment by mengjiao-liu** [comment]
> I'd suggest to add a release note to this PR

Thanks for the suggestion, updated.
```
Pod SecurityContext sysctls name parameter for update requests where the existing object's sysctl contains slashes and kubelet sysctl whitelist support contains slashes.
```
Help to see if it's okay to write like this?

File: pkg/apis/core/validation/validation_test.go
Comment (SergeyKanzhelev) [inline pkg/apis/core/validation/validation_test.go:16503]: would be nice to check with both - `true` and `false`. This will validate that when allow slashes, all the old tests will keep running sucessfully.

File: pkg/api/pod/util.go
Comment (SergeyKanzhelev) [inline pkg/api/pod/util.go:333]: does this comment belong to the function?

File: pkg/apis/core/validation/validation.go
Comment (SergeyKanzhelev) [inline pkg/apis/core/validation/validation.go:3325]:
```suggestion
	// Allow sysctl regex to contain a slash
```

File: pkg/apis/core/validation/validation.go
Comment (SergeyKanzhelev) [inline pkg/apis/core/validation/validation.go:3869]:
```suggestion
	// a sysctl name regex with slash allowed
```

File: pkg/apis/core/validation/validation_test.go
Comment (mengjiao-liu) [reply pkg/apis/core/validation/validation_test.go:16515]: Thanks, Updated.

File: pkg/api/pod/util.go
Comment (mengjiao-liu) [reply pkg/api/pod/util.go]: The description has been modified. Thanks.

File: pkg/apis/core/validation/validation.go
Comment (mengjiao-liu) [reply pkg/apis/core/validation/validation.go]: Updated.

File: pkg/apis/core/validation/validation.go
Comment (mengjiao-liu) [reply pkg/apis/core/validation/validation.go]: Updated.

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

since it's after codefreeze, and will only be part of 1.23, you might need to update comments referring to 1.24 instead of 1.23

**Comment by mengjiao-liu** [comment]
> /lgtm
> 
> since it's after codefreeze, and will only be part of 1.23, you might need to update comments referring to 1.24 instead of 1.23

Thanks for reminding, it has been updated, and request the lgtm label again.

**Comment by mengjiao-liu** [comment]
Update comments referring to 1.24 instead of 1.23.

**Comment by pacoxu** [comment]
/lgtm
after updating the comment to 1.24.

**Comment by mengjiao-liu** [comment]
PTAL @liggitt  @mrunalp

**Comment by mengjiao-liu** [comment]
PTAL @liggitt  @mrunalp

**Comment by mengjiao-liu** [comment]
PTAL @liggitt  @mrunalp

**Comment by mengjiao-liu** [comment]
https://github.com/kubernetes/kubernetes/pull/102393#issuecomment-870696338 
Can someone review this PR? This PR has been around for a long time. Thanks! @mrunalp @dchen1107 @sjenning @Random-Liu

**Comment by mengjiao-liu** [comment]
sig-node approverhave reviewed this PR.
ping @liggitt 
for approval

**Comment by derekwaynecarr** [comment]
@mrunalp can you ensure this doesnt have an impact in cri paths?  it looked totally safe to me.

**Comment by mrunalp** [comment]
I ran a quick test and runc is failing: 
```
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason                  Age   From               Message
  ----     ------                  ----  ----               -------
  Normal   Scheduled               16s   default-scheduler  Successfully assigned default/sysctl-example to 127.0.0.1
  Warning  FailedCreatePodSandBox  15s   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = container create failed: time="2021-10-08T11:46:55-07:00" level=error msg="sysctl \"kernel/shm_rmid_forced\" is not in a separate kernel namespace"
  Warning  FailedCreatePodSandBox  1s    kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = container create failed: time="2021-10-08T11:47:09-07:00" level=error msg="sysctl \"kernel/shm_rmid_forced\" is not in a separate kernel namespace"

```
@kolyshkin ptal

**Comment by mengjiao-liu** [comment]
> We could convert before setting in runc

- I converted sysctl variable before kubelet called runc 
- I have run the test that failed above on k8s environment and the test has passed.

can you review it again？@mrunalp

> but I think runc should probably support what the kernel supports.

I will go to [runc repo](https://github.com/opencontainers/runc/) to open a issue and see how they plan to solve it.

**Comment by AkihiroSuda** [comment]
The runc PR caused regression for values like `net.ipv4.conf.eno2/100.rp_filter`, so maybe we have to revert the runc PR, and just modify the OCI Runtime Spec to prohibit using `/` as an alternative to `.`.

https://github.com/opencontainers/runc/pull/3254#discussion_r739936786
(Thanks to @mengjiao-liu for pointing out)

**Comment by mengjiao-liu** [comment]
> The runc PR caused regression for values like `net.ipv4.conf.eno2/100.rp_filter`, so maybe we have to revert the runc PR, and just modify the OCI Runtime Spec to prohibit using `/` as an alternative to `.`.

But sometimes sysctl variable will have dots and slashes at the same time,
E.g:
"net.ipv4.conf.enp3s0/200.forwarding" or
"net/ipv4/conf/enp3s0.200/forwarding"
They can point to files
/proc/sys/net/ipv4/conf/enp3s0.200/forwarding

If there is no slash, only dots cannot do this.
So we can't disable the slash

And I think the disabled slash is inconsistent with the slash supported by the Linux kernel

ref to https://man7.org/linux/man-pages/man5/sysctl.d.5.html

**Comment by mengjiao-liu** [comment]
Request review again. @mrunalp

**Comment by mengjiao-liu** [comment]
Kubernetes 1.23 Code Freeze is Coming, can there be other sig-node approvers who can help to confirm again? @sig-node-approvers

**Comment by mrunalp** [comment]
@mengjiao-liu Thanks for fixing in runc as well.

/lgtm

**Comment by mrunalp** [comment]
@mengjiao-liu We should update the documentation for this.

**Comment by mengjiao-liu** [comment]
> @mengjiao-liu We should update the documentation for this.

Did you say website document?

**Comment by mrunalp** [comment]
@mengjiao-liu Yes -  https://github.com/kubernetes/website/blob/3ca34b9a3be454055ba234f1d2ff7d55809b5040/content/en/docs/tasks/administer-cluster/sysctl-cluster.md

**Comment by mengjiao-liu** [comment]
Ok, I will open a PR  to update it in the kubernetes/website repository. Thanks for reminding!@mrunalp

**Pull request merged**
